hexkit/protocols/dao.py

```diff
@@ -0,0 +1,415 @@
+        expected_properties = deepcopy(dto_model.schema()["properties"])
+        # (the schema method returns an attribute of the class, making a copy to not
+        # alter the class)
+        del expected_properties[id_field]
+        observed_properties = dto_creation_model.schema()["properties"]
+
+        if observed_properties != expected_properties:
+            raise cls.CreationModelInvalidError()
+
+    @classmethod
+    def _validate_fields_to_index(
+        cls,
+        *,
+        dto_model: type[Dto],
+        fields_to_index: Optional[Sequence[str]],
+    ) -> None:
+        """Checks that all index fields are present in the dto_model. Raises otherwises."""
+
+        if fields_to_index is None:
+            return
+
+        existing_fields = dto_model.schema()["properties"].keys()
+
+        for field in fields_to_index:
+            if field not in existing_fields:
+                raise cls.IndexFieldsInvalidError(f"Field '{field}' not in DTO model.")
+
+    @overload
+    def get_dao(
+        self,
+        *,
+        name: str,
+        dto_model: type[Dto],
+        id_field: str,
+        fields_to_index: Optional[Sequence[str]] = None,
+    ) -> DaoNaturalId[Dto]:
+        ...
+
+    @overload
+    def get_dao(
+        self,
+        *,
+        name: str,
+        dto_model: type[Dto],
+        id_field: str,
+        dto_creation_model: type[DtoCreation],
+        fields_to_index: Optional[Sequence[str]] = None,
+    ) -> DaoSurrogateId[Dto, DtoCreation]:
+        ...
+
+    def get_dao(
+        self,
+        *,
+        name: str,
+        dto_model: type[Dto],
+        id_field: str,
+        dto_creation_model: Optional[type[DtoCreation]] = None,
+        fields_to_index: Optional[Sequence[str]] = None,
+    ) -> Union[DaoSurrogateId[Dto, DtoCreation], DaoNaturalId[Dto]]:
+        """Constructs a DAO for interacting with resources in a database.
+
+        Args:
+            name:
+                The name of the resource type (roughly equivalent to the name of a
+                database table or collection).
+            dto_model:
+                A DTO (Data Transfer Object) model describing the shape of resources.
+            id_field:
+                The name of the field of the `dto_model` that serves as resource ID.
+                (DAO implementation might use this field as primary key.)
+            fields_to_index:
+                Optionally, provide any fields that should be indexed in addition to the
+                `id_field`. Defaults to None.
+            dto_creation_model:
+                An optional DTO model specific for creation of a new resource. This
+                model has to be identical to the `dto_model` except that it has to miss
+                the `id_field`. If specified, resource ID will be generated by the DAO
+                implementation upon. Otherwise (if set to None), resource IDs have to
+                be specified upon resource creation. Defaults to None.
```

Comment:
```suggestion
                implementation upon resource creation. Otherwise (if set to None), resource IDs
                 have to be specified upon resource creation. Defaults to None.
```